crates/ironrdp-cliprdr-format/src/bitmap.rs

```diff
@@ -463,80 +463,83 @@ fn validate_v5_header(header: &BitmapV5Header) -> Result<(), BitmapError> {
-/// and image size:
+/// Computes the stride of an uncompressed RGB bitmap.
+///
+/// INVARIANT: `width <= output (stride) <= width * 4`
+///
+/// In an uncompressed bitmap, the stride is the number of bytes needed to go from the start of one
+/// row of pixels to the start of the next row. The image format defines a minimum stride for an
+/// image. In addition, the graphics hardware might require a larger stride for the surface that
+/// contains the image.
+///
+/// For uncompressed RGB formats, the minimum stride is always the image width in bytes, rounded up
+/// to the nearest DWORD (4 bytes). The following formula is used to calculate the stride:
+///
 /// ```
-/// stride = ((((biWidth * biBitCount) + 31) & ~31) >> 3);
-/// biSizeImage = abs(biHeight) * stride;
+/// stride = ((((width * bit_count) + 31) & ~31) >> 3)
 /// ```
 ///
-/// INVARIANT: bit_count <= 32
-#[allow(clippy::arithmetic_side_effects)]
-fn bmp_stride(width: u16, bit_count: u16) -> usize {
+/// From Microsoft doc: https://learn.microsoft.com/en-us/windows/win32/api/wingdi/ns-wingdi-bitmapinfoheader
+fn rgb_bmp_stride(width: u16, bit_count: u16) -> usize {
     debug_assert!(bit_count <= 32);
-    (((usize::from(width) * usize::from(bit_count)) + 31) & !31) >> 3
+
+    // No side effects, because u16::MAX * 32 + 31 < u16::MAX * u16::MAX < u32::MAX
+    #[allow(clippy::arithmetic_side_effects)]
+    {
+        (((usize::from(width) * usize::from(bit_count)) + 31) & !31) >> 3
+    }
 }
 
-fn transform_bitmap(
+fn bgra_to_top_down_rgba(
     header: &BitmapInfoHeader,
-    input: &[u8],
+    src_bitmap: &[u8],
     preserve_alpha: bool,
-) -> Result<PngEncoderInput, BitmapError> {
-    // If height is positive, DIB is bottom-up, but target PNG format is top-down.
-    let flip = header.flip_vertically();
+) -> Result<PngEncoderContext, BitmapError> {
+    // DIB may be encoded bottom-up, but the format we target, PNG, is top-down.
+    let should_flip_vertically = header.is_bottom_up();
 
     let width = header.width();
     let height = header.height();
 
-    let bit_count = header.bit_count;
+    let src_n_samples = usize::from(header.bit_count / 8);
 
-    let stride = bmp_stride(width, bit_count);
+    let src_stride = rgb_bmp_stride(width, header.bit_count);
 
-    let input_bytes_per_pixel = usize::from(bit_count / 8);
-    let color_type = if preserve_alpha {
-        png::ColorType::Rgba
+    let (dst_color_type, dst_n_samples) = if preserve_alpha {
+        (png::ColorType::Rgba, 4)
     } else {
-        png::ColorType::Rgb
+        (png::ColorType::Rgb, 3)
     };
 
-    let components = color_type.samples();
-    debug_assert!(components <= 4);
-
-    // INVARIANT: height * width * components <= u16::MAX * u16::MAX * 4 < usize::MAX
-    // This is always true because `components <= 4` is checked above, and width & height
-    // bounds are validated on PDU encode/decode
+    // Per invariants: height * width * dst_n_samples <= 10_000 * 10_000 * 4 < u32::MAX
```

Comment:
Bounded by `u32::MAX` instead, because this is what we guarantee in the README.md (assuming usize is either as big as u32 or u64)